build.rs

```diff
@@ -30,44 +30,84 @@ fn rerun_dir<P: AsRef<Path>>(dir: P) {
   }
 }
 
+#[allow(dead_code)]
+fn is_fresh(files: &[&str], out_dir: &str, config: &Path) -> bool {
+  use std::collections::hash_map::DefaultHasher;
+  use std::hash::Hasher;
+  use std::io::Read;
+
+  let mut hasher = DefaultHasher::new();
+
+  let paths = files.iter().map(Path::new).chain(std::iter::once(config));
+
+  for path in paths {
+    if let Ok(mut f) = std::fs::File::open(path) {
+      let mut buf = Vec::new();
+      f.read_to_end(&mut buf).unwrap();
+
+      hasher.write(&buf);
+    } else {
+      panic!("Cannot open {}", path.display());
+    }
+  }
+
+  let hash = hasher.finish().to_be_bytes();
+
+  let hash_path = Path::new(&out_dir).join("asm.hash");
+
+  if let Ok(old_hash) = std::fs::read(&hash_path) {
+    if old_hash == hash {
+      return true;
+    }
+  }
+
+  std::fs::write(hash_path, &hash[..]).unwrap();
```

Comment:
There is a race condition here since the hash file is written before calling nasm.
If nasm is interrupted, future builds will not call nasm again as the hash already matches.
Conversely, if the hash is written after and there is an interruption then nasm will be called again on the next build.

Reply:
I guess the latter condition even if sub-optimal is better than the current status. Thank you for spotting the problem. :)

Reply:
Done :)